File: android/app/src/main/java/com/viant/agently/android/HostedWorkspacePresentation.kt
package com.viant.agently.android

import com.viant.agentlysdk.HostedWorkspaceRestoreState
import com.viant.agentlysdk.ConversationStateResponse
import com.viant.agentlysdk.WorkspaceWindowSnapshot

internal data class HostedWorkspacePresentation(
    val badgeLabel: String,
    val title: String,
    val subtitle: String? = null,
    val supportingText: String = "",
    val navigationIcon: String? = null,
)

internal data class HostedWorkspaceAttachment(
    val turnId: String,
    val presentation: HostedWorkspacePresentation,
)

internal fun isHostedWorkspaceOpeningToolName(value: String): Boolean {
    val normalized = value.trim().lowercase().replace(':', '/')
    return normalized in setOf("ui/view/open", "ui/window/open", "ui/window/show")
}

internal fun hostedWorkspaceAttachment(
    conversationState: ConversationStateResponse?,
    restoreState: HostedWorkspaceRestoreState?,
): HostedWorkspaceAttachment? {
    val presentation = resolveHostedWorkspacePresentation(restoreState) ?: return null
    val turn = conversationState?.conversation?.turns?.asReversed()?.firstOrNull { current ->
        current.execution?.pages.orEmpty().any { page ->
            page.toolSteps.any { step ->
                val status = step.status?.trim()?.lowercase().orEmpty()
                (status.isEmpty() || status in setOf("completed", "succeeded", "success", "done")) &&
                    isHostedWorkspaceOpeningToolName(step.toolName)
            }
        }
    } ?: return null
    return turn.turnId.trim().takeIf(String::isNotEmpty)?.let { HostedWorkspaceAttachment(it, presentation) }
}

internal fun resolveHostedWorkspacePresentation(
    restoreState: HostedWorkspaceRestoreState?
): HostedWorkspacePresentation? {
    restoreState ?: return null
    val window = restoreState.windows.firstOrNull { it.windowId == restoreState.selectedWindowId }
        ?: restoreState.windows.lastOrNull()
    return resolveHostedWorkspacePresentation(window)
}

internal fun resolveHostedWorkspacePresentation(
    window: WorkspaceWindowSnapshot?
): HostedWorkspacePresentation? {
    window ?: return null
    val badgeLabel = normalizeHostedWorkspaceText(window.navigation?.label)
        ?: humanizeHostedWorkspaceKey(window.windowKey)
        ?: "Workspace"
    val normalizedTitle = normalizeHostedWorkspaceText(window.windowTitle)
    val title = when {
        normalizedTitle != null && !normalizedTitle.equals(window.windowKey, ignoreCase = true) -> normalizedTitle
        else -> badgeLabel
    }
    return HostedWorkspacePresentation(
        badgeLabel = badgeLabel,
        title = title,
        subtitle = null,
        supportingText = "Open the ${badgeLabel.lowercase()} workspace.",
        navigationIcon = window.navigation?.icon,
    )
}

private fun humanizeHostedWorkspaceKey(windowKey: String): String? {
    val normalized = windowKey
        .replace("/", " ")
        .replace("_", " ")
        .replace(Regex("([a-z0-9])([A-Z])"), "$1 $2")
        .replace(Regex("([A-Z]+)([A-Z][a-z])"), "$1 $2")
        .trim()
    if (normalized.isEmpty()) {
        return null
    }
    return normalized
        .split(Regex("\\s+"))
        .filter { it.isNotBlank() }
        .joinToString(" ") { token -> humanizeHostedWorkspaceToken(token) }
}

private fun humanizeHostedWorkspaceToken(token: String): String {
    val trimmed = token.trim()
    if (trimmed.isEmpty()) {
        return trimmed
    }
    if (trimmed.all { !it.isLetter() || it.isUpperCase() }) {
        return trimmed
    }
    return trimmed.lowercase().replaceFirstChar { ch ->
        if (ch.isLowerCase()) ch.titlecase() else ch.toString()
    }
}

private fun normalizeHostedWorkspaceText(value: String?): String? {
    val trimmed = value?.trim().orEmpty()
    return trimmed.ifEmpty { null }
}

File: ios/Sources/AgentlyAppFoundation/Workspace/HostedWorkspacePresentation.swift
import Foundation
import AgentlySDK

public struct HostedWorkspacePresentation: Equatable {
    public let badgeLabel: String
    public let badgeSymbolName: String
    public let title: String
    public let subtitle: String?
    public let supportingText: String
}

public struct HostedWorkspaceAttachment: Equatable {
    public let turnID: String
    public let presentation: HostedWorkspacePresentation
}

func isHostedWorkspaceOpeningToolName(_ value: String) -> Bool {
    let normalized = value.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        .replacingOccurrences(of: ":", with: "/")
    return ["ui/view/open", "ui/window/open", "ui/window/show"].contains(normalized)
}

func hostedWorkspaceAttachment(
    conversationState: ConversationStateResponse?,
    restoreState: HostedWorkspaceRestoreState?
) -> HostedWorkspaceAttachment? {
    guard let presentation = resolveHostedWorkspacePresentation(restoreState: restoreState) else { return nil }
    let turn = conversationState?.conversation?.turns.reversed().first(where: { turn in
        turn.execution?.pages.contains(where: { page in
            page.toolSteps.contains(where: { step in
                let status = step.status?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() ?? ""
                return (status.isEmpty || ["completed", "succeeded", "success", "done"].contains(status))
                    && isHostedWorkspaceOpeningToolName(step.toolName)
            })
        }) == true
    })
    guard let turnID = turn?.turnID.trimmingCharacters(in: .whitespacesAndNewlines), !turnID.isEmpty else { return nil }
    return HostedWorkspaceAttachment(turnID: turnID, presentation: presentation)
}

func resolveHostedWorkspacePresentation(
    restoreState: HostedWorkspaceRestoreState?
) -> HostedWorkspacePresentation? {
    guard let restoreState else {
        return nil
    }
    let selectedID = restoreState.selectedWindowId
    let window = restoreState.windows.first(where: { $0.windowId == selectedID }) ?? restoreState.windows.last
    return resolveHostedWorkspacePresentation(window: window)
}

func resolveHostedWorkspacePresentation(
    window: WorkspaceWindowSnapshot?
) -> HostedWorkspacePresentation? {
    guard let window else {
        return nil
    }
    let explicitLabel = normalizeHostedWorkspaceText(window.navigation?.label)
    let badgeLabel = explicitLabel ?? humanizeHostedWorkspaceKey(window.windowKey) ?? "Workspace"
    let normalizedTitle = normalizeHostedWorkspaceText(window.windowTitle)
    let title: String
    if let normalizedTitle,
       normalizedTitle.caseInsensitiveCompare(window.windowKey) != .orderedSame {
        title = normalizedTitle
    } else {
        title = badgeLabel
    }
    return HostedWorkspacePresentation(
        badgeLabel: badgeLabel,
        badgeSymbolName: hostedWorkspaceSymbol(window.navigation?.icon),
        title: title,
        subtitle: nil,
        supportingText: "Open the \(badgeLabel.lowercased()) workspace."
    )
}

private func hostedWorkspaceSymbol(_ value: String?) -> String {
    switch value?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() {
    case "chart": return "chart.xyaxis.line"
    case "document": return "doc.text"
    case "application": return "app"
    default: return "rectangle.topthird.inset.filled"
    }
}

private func humanizeHostedWorkspaceKey(_ key: String) -> String? {
    let normalized = key
        .replacingOccurrences(of: "/", with: " ")
        .replacingOccurrences(of: "_", with: " ")
        .replacingOccurrences(
            of: "([a-z0-9])([A-Z])",
            with: "$1 $2",
            options: .regularExpression
        )
        .replacingOccurrences(
            of: "([A-Z]+)([A-Z][a-z])",
            with: "$1 $2",
            options: .regularExpression
        )
        .trimmingCharacters(in: .whitespacesAndNewlines)
    guard !normalized.isEmpty else {
        return nil
    }
    return normalized
        .split(whereSeparator: \.isWhitespace)
        .map { humanizeHostedWorkspaceToken(String($0)) }
        .joined(separator: " ")
}

private func humanizeHostedWorkspaceToken(_ token: String) -> String {
    let trimmed = token.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !trimmed.isEmpty else {
        return trimmed
    }
    let hasLowercase = trimmed.rangeOfCharacter(from: .lowercaseLetters) != nil
    guard hasLowercase else {
        return trimmed
    }
    let lower = trimmed.lowercased()
    return lower.prefix(1).uppercased() + lower.dropFirst()
}

private func normalizeHostedWorkspaceText(_ value: String?) -> String? {
    let trimmed = value?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
    return trimmed.isEmpty ? nil : trimmed
}
